### PR TITLE
Audio PR - Augmentation support [ Non Silent Region Detection and Slice ]

### PR DESCRIPTION
**Pull request by swetha097**
Add NonSilentRegion and Slice augmentation support in MIVisionX
NOTE : [PR#1323](https://github.com/ROCm/MIVisionX/pull/1323) has to be merged before this PR

### CODE REVIEW

**Review by rrawther — CHANGES_REQUESTED** [review]
Added few comments

**Comment by fiona-gladwin** [comment]
@kiritigowda and @LakshmiKumar23
Please Note RPP PR https://github.com/ROCm/rpp/pull/351 is to be merged before this PR, to prevent build issues

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 1]
```diff
@@ -2558,6 +2558,169 @@ VX_API_ENTRY vx_node VX_API_CALL vxExtRppPreemphasisFilter(vx_graph graph, vx_te
...
+    vx_context context = vxGetContext((vx_reference)graph);
+    if (vxGetStatus((vx_reference)context) == VX_SUCCESS) {
+        vx_uint32 devtype = getGraphAffinity(graph);
+        vx_scalar deviceType = vxCreateScalar(vxGetContext((vx_reference)graph), VX_TYPE_UINT32, &devtype);
+        vx_reference params[] = {
+            (vx_reference)pSrc,
+            (vx_reference)pSrcRoi,
+            (vx_reference)pDst,
+            (vx_reference)pDstRoi,
+            (vx_reference)windowFunction,
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:2573]: it is better to organize some of these parameters into a single user_defined_struct and pass as an array. May be do this as a separate PR.

**Review by rrawther — CHANGES_REQUESTED** [review]
suggest to move all the extra parameters other than tensors into a user_defined strut

File: amd_openvx_extensions/amd_rpp/source/tensor/TensorAddTensor.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    RppPtr_t pSrc2;
+    RppPtr_t pDst;
+    RpptROI *pSrcRoi;
+    RpptROI *pDstRoi;
+    size_t inputTensorDims1[RPP_MAX_TENSOR_DIMS];
+    size_t inputTensorDims2[RPP_MAX_TENSOR_DIMS];
+    vx_enum inTensorType;
+    vx_enum outTensorType;
+};
+
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/TensorAddTensor.cpp:44]: I think tensorAdd is misleading name for this operator. I am assuming it is doing a scalar(const) add to all elements of an  audio sample.  May be, we need to make it as generic like arithmetic (add, subtract, mul etc) and specify how operation works within the parameters.

**Review by rrawther — CHANGES_REQUESTED** [review]
Added comment

**Comment by LakshmiKumar23** [comment]
@SundarRajan28 please sync ToT

**Comment by SundarRajan28** [comment]
@LakshmiKumar23 Synced PR with ToT changes

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h [hunk 3]
```diff
@@ -92,7 +92,8 @@ vx_node createNode(vx_graph graph, vx_enum kernelEnum, vx_reference params[], vx
 vx_status createRPPHandle(vx_node node, vxRppHandle ** pHandle, Rpp32u batchSize, Rpp32u deviceType);
 vx_status releaseRPPHandle(vx_node node, vxRppHandle * handle, Rpp32u deviceType);
 void fillDescriptionPtrfromDims(RpptDescPtr &descPtr, vxTensorLayout layout, size_t *tensorDims);
-void fillAudioDescriptionPtrFromDims(RpptDescPtr &descPtr, size_t *tensorDims, vxTensorLayout layout = vxTensorLayout::VX_NHW);
+void fillGenericDescriptionPtrfromDims(RpptGenericDescPtr &genericDescPtr, vxTensorLayout layout, size_t *maxTensorDims);
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/include/internal_rpp.h:95]: why do we need line 94 & 95? Isn't line 94 for Generic Nodes and 96 for Audio

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 4]
```diff
@@ -2762,6 +2807,28 @@ void fillAudioDescriptionPtrFromDims(RpptDescPtr &descPtr, size_t *maxTensorDims
...
+    genericDescPtr->dims[2] = maxTensorDims[2];
+    genericDescPtr->dims[3] = 1;
+    if(genericDescPtr->dims[2] == 1)
+        genericDescPtr->numDims = 2;
+    else
+        genericDescPtr->numDims = 3;
+    genericDescPtr->strides[0] = genericDescPtr->dims[1] * genericDescPtr->dims[2] * genericDescPtr->dims[3];
+    genericDescPtr->strides[1] = genericDescPtr->dims[2] * genericDescPtr->dims[3];
+    genericDescPtr->strides[2] = genericDescPtr->dims[3];
+    if(tensorLayoutMapping.find(layout) != tensorLayoutMapping.end()) {
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:2825]: Seems redundant from the if condition on line 2811-2812

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h (hunk 3 above)
Comment (rrawther) [reply amd_openvx_extensions/amd_rpp/include/internal_rpp.h:95]: Not sure why there are 3 different functions for fillDescriptions? We need to to converge all into one

File: amd_openvx_extensions/amd_rpp/source/tensor/NonSilentRegionDetection.cpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 3, VX_OUTPUT, VX_TYPE_TENSOR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 4, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 5, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 6, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 7, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 8, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxFinalizeKernel(kernel));
+    }
+    if (status != VX_SUCCESS) {
+    exit:
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/NonSilentRegionDetection.cpp:225]: where is this label used? please remove

File: amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp [hunk 6]
```diff
@@ -0,0 +1,304 @@
+/*
+Copyright (c) 2019 - 2024 Advanced Micro Devices, Inc. All rights reserved.
```
Comment (LakshmiKumar23) [inline amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp:2]: change copyright years

File: amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    // Query the Anchor / Shape Tensor Dims
+    RppSize_t num_dims;
+    size_t anchorTensorDims[RPP_MAX_TENSOR_DIMS];
+    vxQueryTensor((vx_tensor)parameters[4], VX_TENSOR_NUMBER_OF_DIMS, &num_dims, sizeof(num_dims));
+    vxQueryTensor((vx_tensor)parameters[4], VX_TENSOR_DIMS, &anchorTensorDims, sizeof(vx_size) * num_dims);
+    uint total_anchor_dims = 1;
+    for (uint n = 0; n < num_dims; n++)
+        total_anchor_dims *= anchorTensorDims[n];
+
+    total_anchor_dims /= anchorTensorDims[0];
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp:56]: can any of the anchor dims be 0? then there is a possible /0

File: amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 5, VX_INPUT, VX_TYPE_TENSOR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 6, VX_INPUT, VX_TYPE_ARRAY, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 7, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 8, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 9, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxAddParameterToKernel(kernel, 10, VX_INPUT, VX_TYPE_SCALAR, VX_PARAMETER_STATE_REQUIRED));
+        PARAM_ERROR_CHECK(vxFinalizeKernel(kernel));
+    }
+    if (status != VX_SUCCESS) {
+    exit:
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp:298]: I think this label is not used. please remove

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp [hunk 9]
```diff
@@ -2762,6 +2807,28 @@ void fillAudioDescriptionPtrFromDims(RpptDescPtr &descPtr, size_t *maxTensorDims
     }
 }
```
Comment (rrawther) [inline amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:2809]: please combine all 3 functions into 1: Use layout to differentiate

**Review by rrawther — CHANGES_REQUESTED** [review]
please address the review comments

File: amd_openvx_extensions/amd_rpp/include/internal_rpp.h (hunk 3 above)
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/include/internal_rpp.h:95]: Will issue another PR for this . Since it requires changes in all the Audio Files

File: amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp (hunk 7 above)
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/source/tensor/Slice.cpp:56]: The dimension cannot be 0 here

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp (hunk 9 above)
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp:2809]: Will issue this as a separate PR as this would require changes in all the Audio kernel files

File: amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp (hunk 4 above)
Comment (swetha097) [reply amd_openvx_extensions/amd_rpp/source/kernel_rpp.cpp]: modified.

**Comment by swetha097** [comment]
> please address the review comments

Addressed the review comments

**Review by LakshmiKumar23 — APPROVED** [review]

**Review by LakshmiKumar23 — CHANGES_REQUESTED** [review]
@swetha097 why was the exit label added back?

**Review by LakshmiKumar23 — APPROVED** [review]

**Comment by swetha097** [comment]
> @swetha097 why was the exit label added back?

@LakshmiKumar23 The exit label seems to be used in internal_rpp.h file, if removed it leads to build err as its used in one of the checks
```
error: label ‘exit’ used but not defined
53 | #define PARAM_ERROR_CHECK(call){vx_status status = call; if(status!= VX_SUCCESS) goto exit;}
```

**Review by rrawther — APPROVED** [review]
